### PR TITLE
karmada operator: add defaults for karmada cr

### PR DESCRIPTION
**Pull request by calvin0327**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Registry karmada defaults func to scheme.

Test:
we apply a simple karmada cr, like:

```yaml
apiVersion: operator.karmada.io/v1alpha1
kind: Karmada
metadata:
  name: karmada-demo
```

and then the controll manager set defaults for karmada:
image

**Which issue(s) this PR fixes**:
Part of # https://github.com/karmada-io/karmada/issues/2979

**Special notes for your reviewer**:
initialize karmada defaults before entering the logic of reconciliation.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by calvin0327** [comment]
@Poor12 @lonelyCZ PTAL

File: operator/pkg/controller/karmada/controller.go [hunk 1]
```diff
@@ -92,6 +88,31 @@ func (ctrl *Controller) removeFinalizer(ctx context.Context, karmada *operatorv1
 	return controllerruntime.Result{}, nil
 }
 
+func (ctrl *Controller) ensureKarmada(ctx context.Context, karmada *operatorv1alpha1.Karmada) error {
+	// The object is not being deleted, so if it does not have our finalizer,
+	// then lets add the finalizer and update the object. This is equivalent
+	// registering our finalizer.
+	var updated bool
```
Comment (Poor12) [inline operator/pkg/controller/karmada/controller.go:95]: I think we do not need the `updated` temporary variables, even `ContainsFinalizer`.
```
// AddFinalizer accepts an Object and adds the provided finalizer if not present.
// It returns an indication of whether it updated the object's list of finalizers.
func AddFinalizer(o client.Object, finalizer string) (finalizersUpdated bool) {
```
The judgment expression can be directly written as follows:
finalizersUpdated || !reflect.DeepEqual(karmada.Spec, older.Spec)

Comment (calvin0327) [reply operator/pkg/controller/karmada/controller.go:95]: good idea

**Comment by calvin0327** [comment]
@Poor12 others comments?

File: operator/pkg/apis/operator/v1alpha1/defaults.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	hc := obj.Spec.HostCluster
+	if hc.Networking == nil {
+		hc.Networking = &Networking{}
+	}
+	if hc.Networking.DNSDomain == nil {
+		hc.Networking.DNSDomain = pointer.String(constants.KarmadaDefaultDNSDomain)
+	}
+}
+
+func setDefaultsEtcd(obj *KarmadaComponents) {
```
Comment (Poor12) [inline operator/pkg/apis/operator/v1alpha1/defaults.go:73]: Why not set default replicas for etcd?

File: operator/pkg/apis/operator/v1alpha1/register.go [hunk 3]
```diff
@@ -24,19 +22,19 @@ func Resource(resource string) schema.GroupResource {
 }
 
 var (
-	// localSchemeBuilder and AddToScheme will stay in k8s.io/kubernetes.
-	SchemeBuilder      runtime.SchemeBuilder
+	// SchemeBuilder will stay in k8s.io/kubernetes.
+	SchemeBuilder runtime.SchemeBuilder
+	// localSchemeBuilder  will stay in k8s.io/kubernetes.
```
Comment (Poor12) [inline operator/pkg/apis/operator/v1alpha1/register.go:27]: One more space.

File: operator/pkg/apis/operator/v1alpha1/defaults.go (hunk 2 above)
Comment (calvin0327) [reply operator/pkg/apis/operator/v1alpha1/defaults.go:73]: https://github.com/karmada-io/karmada/pull/3349

yes, only if the pr be merged, can the default replicas be set.

Comment (Poor12) [reply operator/pkg/apis/operator/v1alpha1/defaults.go:73]: After #3349 is merged, I think we can set replicas for etcd.

Comment (calvin0327) [reply operator/pkg/apis/operator/v1alpha1/defaults.go:73]: Done. @Poor12

**Comment by lonelyCZ** [comment]
I will test it ASAP.

/assign

**Review by lonelyCZ — COMMENTED** [review]
I just test it in my env, it worked fine!

**Comment by calvin0327** [comment]
@lonelyCZ thanks.

**Pull request merged**